### PR TITLE
CLEAN: Enforce pdep6

### PR DESCRIPTION
**Pull request by MarcoGorelli**
a few bits still need thinking over

### CODE REVIEW

File: pandas/_testing/asserters.py [hunk 1]
```diff
@@ -1443,3 +1445,9 @@ def assert_metadata_equivalent(
             assert val is None
         else:
             assert val == getattr(right, attr, None)
+
+
+def assert_maybe_raises(condition, exception, match):
+    if condition:
+        return pytest.raises(exception, match=match)
+    return contextlib.nullcontext()
```
Comment (MarcoGorelli) [inline pandas/_testing/asserters.py:1453]: this was just a temporary thing i was trying, definitely not suggesting it as part of the api - cleaning it up in the next commit 😇

File: pandas/tests/indexing/multiindex/test_setitem.py [hunk 2]
```diff
@@ -213,13 +213,11 @@ def test_multiindex_assignment_single_dtype(self):
...
-        with tm.assert_produces_warning(
-            FutureWarning, match="item of incompatible dtype"
-        ):
+        with pytest.raises(TypeError, match="Invalid value"):
             df.loc[4, "c"] = arr + 0.5
-        result = df.loc[4, "c"]
-        exp = exp + 0.5
-        tm.assert_series_equal(result, exp)
+        # Upcast so that we can add .5
+        df = df.astype({"c": "float64"})
```
Comment (MarcoGorelli) [inline pandas/tests/indexing/multiindex/test_setitem.py:219]: manually adding this in so we can keep running the rest of the test

File: asv_bench/benchmarks/indexing.py [hunk 3]
```diff
@@ -546,12 +546,9 @@ def time_chained_indexing(self, mode):
 
 
 class Block:
-    params = [
-        (True, "True"),
-        (np.array(True), "np.array(True)"),
-    ]
+    params = [True]
 
-    def setup(self, true_value, mode):
```
Comment (MarcoGorelli) [inline asv_bench/benchmarks/indexing.py:554]: looks like `mode` was already unused in this test?

File: asv_bench/benchmarks/indexing.py [hunk 4]
```diff
@@ -546,12 +546,9 @@ def time_chained_indexing(self, mode):
 
 
 class Block:
-    params = [
-        (True, "True"),
-        (np.array(True), "np.array(True)"),
-    ]
+    params = [True]
```
Comment (MarcoGorelli) [inline asv_bench/benchmarks/indexing.py:549]: `"True"` would now raise, so I'm removing it

Comment (mroeschke) [reply asv_bench/benchmarks/indexing.py:549]: Could you just hardcode `True` in the only benchmark where its used?

File: pandas/core/internals/blocks.py [hunk 5]
```diff
@@ -428,7 +428,7 @@ def split_and_operate(self, func, *args, **kwargs) -> list[Block]:
     # Up/Down-casting
 
     @final
-    def coerce_to_target_dtype(self, other, warn_on_upcast: bool = False) -> Block:
+    def coerce_to_target_dtype(self, other, raise_on_upcast: bool = False) -> Block:
```
Comment (mroeschke) [inline pandas/core/internals/blocks.py:431]: Are there any usages of this method, outside this file, that does not set this to `raise_on_upcast=False`?

Comment (MarcoGorelli) [reply pandas/core/internals/blocks.py]: there's no usage of this method at all outside this file

i've removed the default `=False` anyway, better to be explicit about this one

**Review by mroeschke — COMMENTED** [review]
Code changes look good. Could you add something in the whatsnew (section or one-liner, up to you) about this enforcement?

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @MarcoGorelli

**Pull request merged**